### ISSUE TITLE
Get Hashstore Conversion Working in k8s

### ISSUE THREAD

**Issue by taojing2002** [NCEAS/metacat#1964]
We modified the code in the `MetacatInitializer.contextInitialized` method to make sure it will not call db in k8s. But we haven't tested it.

**Comment by artntek** [comment]
### NOTE: I left this note for myself in August. Jing might have changed the code/fixed the problem since then? Including here for reference, just in case:

Tested on k8s in August 2024 (brooke namespace on dev cluster): 

This happens before my k8s initializer upgrades the DB:
```shell
metacat 20240827-21:10:09: [ERROR]: Could not determine if metacat is configured due to the 
admin exception: Cannot get the status from the db for version 3.1.0 since ERROR: 
relation "version_history" does not exist
Position: 36 [edu.ucsb.nceas.metacat.util.ConfigurationUtil:isMetacatConfigured:50]
```
because of the new line in `ConfigurationUtil::isMetacatConfigured` :
```java
  && HashStoreConversionAdmin.isConverted();
```
that method calls `HashStoreConversionAdmin.java::getStatus()`, which tries to query the DB

_(MB NOTE to self there are TGZ backup files of metacat files and postgres files in /var/metacat and /bitnami/postgresql)_

**Comment by taojing2002** [comment]
I didn't test it yet. Also the MetacatInitializerIT class fails for k8s in
the develop branch. We need to fix it.

**Comment by artntek** [comment]
Tested on dev cluster using a copy of the data and DB from `mn-sandbox-ucsb-2`

    - used Metacat 3.1.0-SNAPSHOT, built from `develop`.
    - 33 minutes to convert 36GB (i.e. ~1GB/min). 
    - Total [57297 objects](https://metacat-dev.test.dataone.org/metacat/d1/mn/v2/object)
    - error output:
        - `generalError_2024-10-28_23-05-38.txt` -- 1028 entries
        - `nonMatchingChecksum_2024-10-28_23-05-38.txt` -- 2 entries
        - `noSuchAlgorithm_2024-10-28_23-05-38.txt` -- 90 entries
    - tot ~1.9% error rate

**Comment by artntek** [comment]
## Conversion Errors: Summary and Action Items

### Discussed the above Error messages in our [10/29/24 backend meeting](https://hpad.dataone.org/4dNnT9mrRNyq0AgkOfjtPQ?both#2024-10-29).

#### `generalError_2024-10-28_23-05-38.txt` -- 1028 entries

- Probably due to file permissions, e.g.:
  ```shell

  ```

- Probably due to bad "test" datasets, e.g.:
  ```shell
  testListViewsExist_20167494910441 Given file path: /var/metacat/documents/autogen.2016031409493873334.1 does not exist.
  ```
    - hypothesis is that we need group RW to do hard links. Test this
    - need consistent ownership and permissions across all the cephfs data under ceph/repos/... See this comment below.

  ```shell
  testUpdateSystemMetadata_NoSerialVersion_20171372385360 Cannot invoke "String.length()" because "<parameter1>" is null
  ``` 
    - re-run for GOA corpus (after fixing permissions) - should be "clean"

#### `nonMatchingChecksum_2024-10-28_23-05-38.txt` -- 2 entries
- these are likely old datasets that predate stricter validation. If so, need to investigate and fix by hand. Shouldn't be many
- if within last decade, we should investigate how - this shouldn't be possible with metacat api

#### `noSuchAlgorithm_2024-10-28_23-05-38.txt` -- 90 entries
- before dataONE, we weren't stricty about algo names. Jing & Jeanette fixed a bunch of these in the past - there shouldn't be any on prod. We can check before doing the hashstore conversion, using a SQL query. Should be old ticket(s) that might help. If algo is one we don't support (rather then sha1 vs. sha-1), need to discuss further.

**Comment by artntek** [comment]
## Need consistent ownership and permissions across all the cephfs data under `ceph/repos/*`
* UID & GID - coordinate with Nick
* Standardize on posix file permissions (some currently 660, some 640)
* NOTE: metadig & indexer also need read-only access to `metacat/hashstore` - need to coordinate if ownership changes for existing deployments!

**Comment by artntek** [comment]
Realized that when hashstore writes files. it does not grant `read` access to the `group`. Files are readable only by the `user` (`tomcat`, in our case). This is a problem for other clients like the indexer and metadig, for example. 

Verified that indexing is broken in 3.1.0 on k8s (it works on localhost/legacy because indexer and metacat both run in the same tomcat instance as the same user, but on k8s, only the group is the same).

**Comment by artntek** [comment]
11/11/24 at ~4pm: Started running the conversion on the ceph copy of the prod ADC corpus.

Started with:
- a completely empty `/var/metacat/hashstore/` directory
- empty `checksums` DB table
- `systemmetadata.checksum_algorithm` fixed to contain only "official" names, thus avoiding noSuchAlgorithm errors, and
- ownership & permissions fixed so converter can access the data files on disk

https://arcticdata.io/metacat/d1/mn/v2/object shows total="1115730", and converter creates 5 checksums per object, so checksums table should ultimately contain of the order of 5.5M rows, not accounting for errors.
```shell
kubectl exec metacatarctic-postgresql-0 -- \
    bash -c "psql -U metacat -c 'select count(*) from checksums'"
```

**Comment by artntek** [comment]
## Conversion Errors encountered and explained

1. Data object already exists
   ```logs
   [WARN]: Data object already exists at: /var/metacat/hashstore/objects/a8/ce/e0/67f87732702a293772d070319ca8c8b35dceed351032b1040de85059c6
   [org.dataone.hashstore.hashstoreconverter.FileHashStoreLinks:storeHardLink:121]
   ```
   That warning should be okay - it means that HashStore attempted to create a hard link but
   it already exists (where 2 datasets point to the same data object, for example). Since 
   this is "normal", we agreed that  `[WARN]` isn't appropriate, so @doulikecookiedough is 
   changing the code to log this as a lower level.

2. "Checksum supplied does not match what has been calculated"
   ```
   metacat 20241112-17:06:23: [ERROR]: Checksum supplied:
   2a3f11c7c704e0128bfd3829da1ebe5c1bd7957cb8a01de5dbc2b9d097c022b6 does not match what 
   has been calculated: 447d5f617a6546a892f9b5e0ec1dba6763fc11919f64c4d1e3361aaa4f8a4aee for 
   pid: urn:uuid:56e62d5e-76f9-4694-bd88-2a4931845ebb and checksum algorithm: SHA-256 
   [org.dataone.hashstore.hashstoreconverter.FileHashStoreLinks:storeHardLink:103]
   ```
   This means the checksum in the system metadata doesn't match the checksum of the actual
   object. We decided to change the converter code to update the system metadata and log
   additional info so the data remains available, and the operator can pursue resolution
   with the dataset owner. See #2013 

3.

**Comment by artntek** [comment]
### 11/18/24: Set `storage_upgrade_status` to `pending` to retry conversion with failed objects, and new metacat image with the above fixes.

Errors in Log:

1. (KNOWN, RESOLVED) "Checksum supplied [...] does not match what has been calculated"

```shell
metacat 20241119-01:43:38: [ERROR]: Checksum supplied: 8b38544356f6d171f43cf77d5dfb444e401698c2 does not match what has been calculated: 54d92946da3c901bc4da65e86cc98d2e for pid: urn:uuid:c98bd99d-cb0c-4e22-97af-af401da44447 and checksum algorithm: MD5 [org.dataone.hashstore.hashstoreconverter.FileHashStoreLinks:storeHardLink:103]
```

The above is a precursor to this (so there are 2 ERROR lines for each of these)
```
metacat 20241119-01:53:10: [ERROR]: Object: urn:uuid:5d815400-9247-4276-851d-0630da803094 was converted successfully, but original MD5 checksum (c66dfd8d6e67fd2a470dfc10c51f75274907a155) was incorrect. System metadata was updated with new checksum (8425656bd85ac0ec2242c22c16da1b06). Old object for reference: /var/metacat/data/autogen.2020060514262434876.1 
```
...and SOME of the above have this addendum (concatenated to the same log line):
```
Note: the original checksum would have been correct if the SHA-1 algorithm was used. Was the wrong algorithm recorded? [edu.ucsb.nceas.metacat.admin.upgrade.HashStoreUpgrader:reConvertUnMatchedChecksumObject:629]
```
Checksum in the system metadata doesn't match the checksum of the actual
object. Converter code new updates the system metadata and logs
additional info so the data remains available, and the operator can pursue resolution
with the dataset owner.


2. (KNOWN, RESOLVED) Cannot move the object [...] since Docid not found in the identifier table

```
metacat 20241119-01:20:36: [ERROR]: Cannot move the object urn:uuid:2de418e3-d9bb-4b7f-82af-ef5885da6b9b to hashstore since Docid not found in the identifier table: urn:uuid:2de418e3-d9bb-4b7f-82af-ef5885da6b9b [edu.ucsb.nceas.metacat.admin.upgrade.HashStoreUpgrader:upgrade:287]
```
Needs manual intervention

3. Can't find the object [...] in the Metacat legacy store.

```
metacat 20241119-01:29:20: [ERROR]: Cannot move the object urn:uuid:fc690663-8a87-4ea7-99e1-11c01563885e to hashstore since Can't find the object /var/metacat/data/autogen.2024110616562124397.1 in the Metacat legacy store. [edu.ucsb.nceas.metacat.admin.upgrade.HashStoreUpgrader:upgrade:287]
```

**Comment by taojing2002** [comment]
Regarding item 3, the object is in the arcticdata.io:
```
***@***.***:/home/tao# ls -al
/var/metacat/data/autogen.2024110616562124397.1
-rw-r----- 1 tomcat tomcat 268510 Nov  6 16:56
/var/metacat/data/autogen.2024110616562124397.1
```
So it is pretty new. Did you rsynic to k8s?

**Comment by mbjones** [comment]
Looking at your single missing object logs by removing the `event='delete'` constraint, I see that there was an `upload` event from the Metacat API by a researcher on 2016-11-11, and a `updateSystemMetadata` event using our RClient tools by another of our past interns/coordinators (Dominic) on 2019-02-25.

Looking at the system metadata table, I see a record still exists for that file: 

```
metacat=# select * from systemmetadata where guid like 'arctic-data.9767.1';
        guid        | series_id | serial_version |    date_uploaded    |            rights_holder             |             checksum             | checksum_algorithm | origin_member_node | authoritive_member_node |      date_modified      |              submitter               | object_format | size | archived | replication_allowed | number_replicas | obsoletes | obsoleted_by | media_type |       file_name
--------------------+-----------+----------------+---------------------+--------------------------------------+----------------------------------+--------------------+--------------------+-------------------------+-------------------------+--------------------------------------+---------------+------+----------+---------------------+-----------------+-----------+--------------+------------+------------------------
 arctic-data.9767.1 |           | 1              | 2016-11-11 16:00:00 | http://orcid.org/0000-0001-9790-3510 | 6a2dbd5a94409a8db54862346aaa729c | MD5                | urn:node:ARCTIC    | urn:node:ARCTIC         | 2019-02-25 16:15:33.047 | http://orcid.org/0000-0001-9790-3510 | text/csv      | 2016 | f        | f                   |               0 |           |              |            | arctic-data.9767.1.csv
```

So it's a CSV file named `arctic-data.9767.1.csv` and we have the checksum for it, and. it was created on the same day in the access log.  

Running a SOLR query (`curl -s "https://arcticdata.io/metacat/d1/mn/v2/query/solr/?q=id:"arctic-data.9767.1"&rows=10&wt=json" |jq .`) against the id shows the details for the file, which are:

```json
{
  "responseHeader": {
    "status": 0,
    "QTime": 0,
    "params": {
      "q": "id:arctic-data.9767.1",
      "fq": "(readPermission:\"public\")OR(writePermission:\"public\")OR(changePermission:\"public\")OR(isPublic:true)",
      "rows": "10",
      "wt": "javabin",
      "version": "2"
    }
  },
  "response": {
    "numFound": 1,
    "start": 0,
    "numFoundExact": true,
    "docs": [
      {
        "id": "arctic-data.9767.1",
        "sku": "arctic-data.9767.1",
        "identifier": "arctic-data.9767.1",
        "fileName": "arctic-data.9767.1.csv",
        "formatId": "text/csv",
        "formatType": "DATA",
        "size": 2016,
        "checksum": "6a2dbd5a94409a8db54862346aaa729c",
        "submitter": "http://orcid.org/0000-0001-9790-3510",
        "checksumAlgorithm": "MD5",
        "rightsHolder": "http://orcid.org/0000-0001-9790-3510",
        "replicationAllowed": false,
        "numberReplicas": 0,
        "archived": false,
        "dateUploaded": "2016-11-12T00:00:00Z",
        "updateDate": "2016-11-12T00:00:00Z",
        "dateModified": "2019-02-26T00:15:33.047Z",
        "datasource": "urn:node:ARCTIC",
        "authoritativeMN": "urn:node:ARCTIC",
        "isPublic": true,
        "dataUrl": "https://cn.dataone.org/cn/v2/resolve/arctic-data.9767.1",
        "isService": false,
        "serviceCoupling": "false",
        "synchronization_failed_count_i": 7,
        "updateSystemMetadata_count_i": 1,
        "changePermission": [
          "http://orcid.org/0000-0001-9790-3510",
          "CN=arctic-data-admins,DC=dataone,DC=org"
        ],
        "readPermission": [
          "public",
          "http://orcid.org/0000-0001-9790-3510",
          "CN=arctic-data-admins,DC=dataone,DC=org"
        ],
        "resourceMap": [
          "resourceMap_arctic-data.9785.1"
        ],
        "isDocumentedBy": [
          "arctic-data.9785.1"
        ],
        "writePermission": [
          "http://orcid.org/0000-0001-9790-3510",
          "CN=arctic-data-admins,DC=dataone,DC=org"
        ],
        "language": "",
        "_version_": 1666538750359896065
      }
    ]
  }
}
```

So now we know it is associate with the package with resource map `resourceMap_arctic-data.9785.1` and is documented by the metadata file `arctic-data.9785.1` Viewing that metadata file at https://arcticdata.io/catalog/view/arctic-data.9785.1 (you have to be logged in as a curator because it is private) shows a lot more metadata about the file. 

With all of that information, maybe Nick can help track it down in a backup, or there would be more info to work from to maybe find the file on our online systems?

**Comment by mbjones** [comment]
Nick says he may have found a backup tape to restore from. However, looking more carefully, I see a couple of things:

- The data package with metadata identifier `arctic-data.9785.1` is marked as **archived**, and its access control is set to **private**. So, nobody except the datateam or the originator can access the dataset and its files anyways
- I find a newer version of a dataset that is published with a DOI a few days later, and seems to have a very similar fileset. The newer dataset also includes a `Cherski_river_discharge.csv` file with a different PID [urn:uuid:3f9f174f-f0db-4cb1-8fef-ddf99f8bb2c1](https://arcticdata.io/catalog/view/doi%3A10.18739%2FA26Q1SH4T#urn%3Auuid%3A3f9f174f-f0db-4cb1-8fef-ddf99f8bb2c1) and a different checksum. 

Because the latter dataset is public with a DOI, I suspect what happened is that the original upload was deemed invalid or wrong, and so the package was archived, and a new one created to replace it. The old package was made private so that it would not be accessible. I'm not sure why for sure, but I'm pretty sure it was a mistaken duplicate. I found a [notes file from our data team processing folder](https://github.nceas.ucsb.edu/KNB/arctic-data/blob/master/datateam/data-processing/2016-Shiklomanov/archiving%20duplicate%20datasets) that contained the following reference indicating that the package in question `arctic-data.9785.1` was being archived as a duplicate:

```
#Archive duplicate ADC datasets, keep arctic-data.9727.1

archive(mn, "arctic-data.9787.1")
archive(mn, "arctic-data.9785.1")
archive(mn, "arctic-data.9784.1")
archive(mn, "arctic-data.9752.1")
archive(mn, "arctic-data.9743.1")
archive(mn, "arctic-data.9734.1")
archive(mn, "arctic-data.9733.1")
archive(mn, "arctic-data.9723.1")
archive(mn, "arctic-data.9674.1")
```

In conclusion, I think we should write this off as an intentionally deleted file, possibly mark it as such in the access log (lets discuss), and call it good.

**Comment by taojing2002** [comment]
We have a plenty duplicate files with the same checksum in our system:
```
select guid from systemmetadata where checksum='6a2dbd5a94409a8db54862346aaa729c' and checksum_algorithm='MD5';
        guid        
--------------------
 arctic-data.9770.1
 arctic-data.9705.1
 arctic-data.9738.1
 arctic-data.9706.1
 arctic-data.9720.1
 arctic-data.9721.1
 arctic-data.9747.1
 arctic-data.9765.1
 arctic-data.9766.1
 arctic-data.9767.1
```

**Comment by taojing2002** [comment]
Matthew and I copied the file `arctic-data.9766.1` to regenerated `arctic-data.9767.1`. And we made sure that they both have the checksum `6a2dbd5a94409a8db54862346aaa729c`
